### PR TITLE
Auto-expire past open duty swap requests via scheduled cronjob

### PR DESCRIPTION
**Pull request by pietbarber**
## Summary
This PR implements automatic expiry of stale duty swap requests whose duty date has already passed.

Previously, requests could remain `open` indefinitely after the original date elapsed, leaving stale records and causing confusing member communication when old requests were later handled manually.

This change adds a scheduled backend workflow that closes those stale requests in a consistent, auditable way.

## Problem
- `DutySwapRequest` entries with `status="open"` and `original_date < today` were not automatically closed.
- Pending offers on those stale requests also remained pending.
- Lifecycle state could drift from real-world duty operations.

### Backend cron command
- Added `duty_roster/management/commands/expire_past_swap_requests.py`.
- Uses `BaseCronJobCommand` for distributed-lock-safe execution in multi-pod environments.
- Finds stale open requests and transitions them to `expired`.
- Auto-declines related pending offers (`auto_declined`) and sets `responded_at`.
- Supports `--dry-run` behavior with clear logging.

### Notification flow
- Added `send_request_expired_notifications` integration in `duty_roster/views_swap.py`.
- Added dedicated templates:
  - `duty_roster/templates/duty_roster/emails/swap_request_expired_requester.html`
  - `duty_roster/templates/duty_roster/emails/swap_request_expired_offerer.html`

### Infrastructure / scheduling
- Added scheduled job entry to `k8s-cronjobs.yaml` so expiry runs automatically.

### Documentation
- Updated architecture/workflow/user-facing docs and issue resolution docs:
  - `docs/cronjob-architecture.md`
  - `docs/duty-swap-user-guide.md`
  - `docs/workflows/duty-swap-workflow.md`
  - `docs/resolved-issues/README.md`
  - `docs/resolved-issues/issue-933-duty-swap-auto-expiry.md`

## Tests
Focused validation run on this branch:

```bash
pytest duty_roster/tests/test_duty_swap.py -k SwapRequestExpiryCronjob -q
```

Result: **7 passed**

Covered scenarios include:
- Expiring past-dated open requests.
- Auto-declining related pending offers.
- No mutations during dry-run mode.
- Ensuring future/open and non-open historical requests remain unchanged.

## Operational Notes
- No schema migrations required.
- Command runtime and lock semantics align with existing cronjob infrastructure.
- Active deadline/schedule is configured via Kubernetes cronjob manifest.

Closes #933

### CODE REVIEW

File: duty_roster/tests/test_duty_swap.py [hunk 1]
```diff
@@ -1759,6 +1760,136 @@ def test_opportunity_appears_for_qualified_member(self, client, alice, site_conf
         role_labels = {o["role_label"] for o in matching}
         assert any("tow" in label.lower() for label in role_labels)
 
+
+@pytest.mark.django_db
+class TestSwapRequestExpiryCronjob:
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply duty_roster/tests/test_duty_swap.py:1767]: Reorganized tests in duty_roster/tests/test_duty_swap.py so volunteer-opportunity edge-case tests are no longer under TestSwapRequestExpiryCronjob. Added a dedicated volunteer edge-case class, and kept the cronjob class focused on expiry behavior.

File: duty_roster/tests/test_duty_swap.py [hunk 2]
```diff
@@ -1759,6 +1760,136 @@ def test_opportunity_appears_for_qualified_member(self, client, alice, site_conf
...
+        )
+
+        notified = []
+
+        def _mock_notify(request_obj, auto_declined_offers=None):
+            notified.append((request_obj.pk, len(auto_declined_offers or [])))
+
+        monkeypatch.setattr(
+            "duty_roster.management.commands.expire_past_swap_requests.send_request_expired_notifications",
+            _mock_notify,
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply duty_roster/tests/test_duty_swap.py:1794]: Added focused rendering/mail coverage in duty_roster/tests/test_duty_swap.py::test_expiry_notifications_render_and_send_requester_and_offerer_templates. This exercises send_request_expired_notifications directly, verifies both emails are generated, and asserts rendered HTML content from requester and offerer templates.

File: docs/workflows/duty-swap-workflow.md [hunk 3]
```diff
@@ -187,13 +187,20 @@ When someone offers a **swap** (not a cover), the system checks if the proposed
...
 - Requester cancels request → **all originally notified members** get "all clear" email (whether they made an offer or not)
+- Nightly system job expires stale requests (`open` with duty date in the past) → request marked `expired`, pending offers marked `auto_declined`, and expiry notifications sent
 - No offers by deadline → Duty Officer decides based on role:
   - **Critical roles (Tow Pilot, Duty Officer)**: Operations MUST be cancelled
     - No tow pilot = can't fly
     - No duty officer = no one in charge
   - **Optional roles (Instructor, ADO)**: Operations can proceed without role, or DO can cancel, or DO can manually assign someone
 - Request stays open until resolved or duty day arrives
 
+### 8. Automatic Expiry of Stale Requests
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply docs/workflows/duty-swap-workflow.md]: Renumbered sections in docs/workflows/duty-swap-workflow.md to preserve ordered flow: Automatic Expiry is now section 6, Roles Covered is 7, and Site Configuration Constraints is 8.

File: docs/cronjob-architecture.md [hunk 4]
```diff
@@ -100,6 +100,7 @@ Uses PostgreSQL's atomic operations:
    - ✅ `send_duty_preop_emails.py` - **SCHEDULED DAILY**
    - ✅ `send_maintenance_digest.py` - **SCHEDULED WEEKLY**
    - ✅ `expire_ad_hoc_days.py` - **SCHEDULED DAILY**
+    - ✅ `expire_past_swap_requests.py` - **SCHEDULED DAILY** (Issue #933 stale swap request expiry)
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply docs/cronjob-architecture.md]: Fixed the bullet indentation in docs/cronjob-architecture.md so expire_past_swap_requests is aligned with sibling command bullets and renders at the same list level.

File: docs/resolved-issues/issue-933-duty-swap-auto-expiry.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+- Pending offers become auto-declined.
+- Dry-run mode produces no mutations.
+- Future open requests and non-open historical requests remain unchanged.
+
+Validation run:
+- `pytest duty_roster/tests/test_duty_swap.py -k SwapRequestExpiryCronjob -q`
+- Result: passing.
+
+### Performance Impact
+The command executes a filtered, indexed date/status query and processes only stale open requests. Runtime overhead is low and bounded by stale-request volume, with a 5-minute active deadline in Kubernetes.
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply docs/resolved-issues/issue-933-duty-swap-auto-expiry.md]: Updated docs/resolved-issues/issue-933-duty-swap-auto-expiry.md to remove the inaccurate indexed-query claim. The section now states it uses a filtered date/status query without implying an index that does not exist.

File: docs/duty-swap-user-guide.md [hunk 6]
```diff
@@ -192,6 +199,7 @@ You'll receive emails for:
 - **Offers accepted** by requesters
 - **Offers declined** by requesters
 - **Requests cancelled** by requesters
+- **Requests expired** automatically after the duty date has passed
 - **Emergency escalations** (if you're Duty Officer)
 
 All emails include direct links to the relevant pages.
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply docs/duty-swap-user-guide.md]: Updated docs/duty-swap-user-guide.md to qualify the email-link statement: most emails include direct links, while expiry notices provide summary context and direct users back to Duty Swap history.

File: docs/cronjob-architecture.md [hunk 7]
```diff
@@ -99,7 +99,8 @@ Uses PostgreSQL's atomic operations:
 1. ✅ Converted existing commands to use base class:
    - ✅ `send_duty_preop_emails.py` - **SCHEDULED DAILY**
    - ✅ `send_maintenance_digest.py` - **SCHEDULED WEEKLY**
-   - ✅ `expire_ad_hoc_days.py` - **SCHEDULED DAILY**
+    - ✅ `expire_ad_hoc_days.py` - **SCHEDULED DAILY**
+    - ✅ `expire_past_swap_requests.py` - **SCHEDULED DAILY** (Issue #933 stale swap request expiry)
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply docs/cronjob-architecture.md]: Aligned both command bullets in docs/cronjob-architecture.md to the same indentation level as the sibling list items under the converted-command section, so Markdown renders them as peer bullets instead of nested content.

File: duty_roster/management/commands/expire_past_swap_requests.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        "Expire open duty swap requests for dates in the past and auto-decline "
+        "their pending offers"
+    )
+    job_name = "expire_past_swap_requests"
+    max_execution_time = timedelta(
+        minutes=5
+    )  # Matches K8s CronJob activeDeadlineSeconds=300
+
+    def execute_job(self, *args, **options):
+        today = timezone.localdate()
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply duty_roster/management/commands/expire_past_swap_requests.py]: Updated duty_roster/management/commands/expire_past_swap_requests.py to use timezone.now().date() for the cutoff date, aligning expiry behavior with the CronJob's UTC schedule window. Added regression coverage in duty_roster/tests/test_duty_swap.py (test_uses_utc_cutoff_date_for_nightly_run) to assert a request from the prior UTC day is expired during the 03:10 UTC run.

File: duty_roster/management/commands/expire_past_swap_requests.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+                for offer in pending_offers:
+                    offer.status = "auto_declined"
+                    offer.responded_at = now_ts
+                    offer.save(update_fields=["status", "responded_at"])
+
+                send_request_expired_notifications(
+                    swap_request,
+                    auto_declined_offers=pending_offers,
+                )
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply duty_roster/management/commands/expire_past_swap_requests.py]: Updated duty_roster/management/commands/expire_past_swap_requests.py to defer expiry emails with transaction.on_commit(...), so notification work runs only after a successful commit and no longer holds row locks during template/SMTP processing. Added regression coverage in duty_roster/tests/test_duty_swap.py with test_expiry_notifications_are_deferred_until_transaction_commit, and kept targeted cronjob tests passing.

File: docs/cronjob-architecture.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (pietbarber) [reply docs/cronjob-architecture.md]: Normalized indentation in docs/cronjob-architecture.md so both expire_* bullets now use the same leading-space depth as sibling command bullets in the converted-command list, ensuring they render as peer list items.
